### PR TITLE
Color picker – Add full gradient support

### LINKED ISSUE #263: Color Picker - Gradient Editing
As a user I want to be able to set color gradients

This ticket is a followup to #106 

This ticket has a followup in #403 which deals with visually displaying and manipulating the gradient shape and color stops on the selected element/page background.

This leaves this ticket dealing with **only** the color picker and gradient manipulation inside the color picker.

## Feature brief
For this we need a way to store a pattern in data, where the Pattern can be either opaque or some gradient.

After a lot of discussion with @swissspidy and @dvoytenko, we've decided to use a data format throughout the application, but to make sure it's as compact as possible for the common case of just an opaque color. This data format will then be assigned to anything that can have a user-configurable color (whether or not it can be a gradient or not) and passed to the color picker.

Helper functions will let components get the background CSS value from a Pattern object.

### Format
```
Percentage: float(0..1)
Dimension: int(0..255)
Color: object(
	r: Dimension,
	g: Dimension,
	b: Dimension,
	a: Percentage(default: 1),
)
Stop: object(
	color: Color,
	stop: Percentage,
)
PatternType: string('solid','linear','radial','conic')
Pattern: object(
	type: PatternType(default: 'solid'),
        // the following attribute is only used for type=opaque
	color: Color,
        // the following attribute is only used for type=linear,radial,conic
	stops: array(Stop),
        // the following attribute is only used for type=linear,conic
	rotation: Percentage(default: 0),
        // the following attribute is only used for type=radial,conic
	center: array(Percentage)[2](default: [.5,.5]),
        // the following attribute is only used for type=radial
	size: array(Percentage)[2](default: [.5,.5]),
)
```

### Examples:
```js
// transparent
e.background = null;
e.background = {
  type: 'transparent',
};
// converted to a CSS property:
//   background-color: transparent;

// pure blue - note that the type is excluded as 'solid' is the default
e.background = {
  color: {r:0,g:0,b:255},
};
// converted to a CSS property:
//   background-color: #00f;

// simple linear gradient from red to blue
e.background = {
  type:'linear',
  stops: [
    {color:{r:255,g:0,b:0}, stop: 0},
    {color:{r:0,g:0,b:255}, stop: 1}
  ],
};
// converted to a CSS property:
//   background-image: linear-gradient(#f00, #00f);

// conic gradient with 3 stops starting at .2 rotation (72 degrees)
e.background = {
  type:'conic',
  stops: [
    {color:{r:255,g:0,b:0},stop:.1},
    {color:{r:0,g:255,b:0},stop:.3},
    {color:{r:0,g:0,b:255}, stop: 1}
  ],
  rotation: .2,
};
// converted to a CSS property:
//   background-image: conic-gradient(from .2turn, #f00 .1turn, #0f0 .3turn, #00f);

// elongated radial gradient in the top left corner with 4 stops starting at transparent
e.background = {
  type:'radial',
  stops: [
    {color:{r:255,g:0,b:0,a:0},stop:.1},
    {color:{r:255,g:0,b:0},stop:.3},
    {color:{r:0,g:255,b:0},stop:.3},
    {color:{r:0,g:0,b:255}, stop: 1},
  ],
  size:[.5,.2],
  center:[0,0],
};
// converted to a CSS property:
//   background-image: radial-gradient(ellipse 50% 20% at 0% 0%, rgba(255,0,0,0) 10%, #f00 30%, #0f0 30%, #00f 100%);
```

### Todo
Before considering adding gradient support to the color picker:

 - [x] Create format in JSDoc (including gradients)
 - [x] Create helper functions converting to CSS (including gradients)
 - [x] Modify existing color picker to use this input and output format
 - [x] Configure color picker to allow transparent and/or gradient where applicable
 - [x] Create migration for all existing colors to this format
 - [x] Replace any color selector with color picker in design panels
 - [x] Use helper functions in display and output components

Then implementing gradients is "simply" adding the desired functionality to the color picker and the rest of the functionality should follow:
- [x] Create internal reducer
- [ ] Unit test reducer
- [x] Add color type picker (solid/linear/radial/conic)
- [x] Add new gradient stop component
- [x] Add stop reverse button
- [x] Add gradient 90° rotation button

## Acceptance Criteria
1. The user is able to set linear color gradients
2. The user is able to set radial color gradients
3. The user is able to set color stops
4. The user is able to reverse color stops
5. The user is able to rotate the gradient direction 90 degrees clockwise

### PR DESCRIPTION
**Pull request by barklund**
Full gradient support added in this PR.

With full test coverage for color picker, color input and all color-related util functions!

See it in action here: https://youtu.be/V3qsoONXUwk

Fully fixes #263.

### CODE REVIEW

File: assets/src/edit-story/components/colorPicker/test/insertStop.js [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+ */
+import insertStop from '../insertStop';
+
+const WHITE = { r: 255, g: 255, b: 255 };
+const BLACK = { r: 0, g: 0, b: 0 };
+const ABLACK = { r: 0, g: 0, b: 0, a: 0 };
+const GRAY80 = { r: 51, g: 51, b: 51 };
+const MGRAY80 = { r: 51, g: 51, b: 51, a: 0.2 };
+
+expect.extend({ toBeDeepCloseTo });
```
Comment (swissspidy) [inline assets/src/edit-story/components/colorPicker/test/insertStop.js:33]: Could also move this to `jest.setup.js` where other matchers are added

Comment (barklund) [reply assets/src/edit-story/components/colorPicker/test/insertStop.js:33]: I didn't really know if it was generic enough for that. Let's leave it here until we need it elsewhere.

Comment (barklund) [reply assets/src/edit-story/components/colorPicker/test/insertStop.js:33]: It's also a [tiny plugin with only 13 stars](https://github.com/maasencioh/jest-matcher-deep-close-to) (but MIT licensed), so I didn't want to put it in core.

File: tests/phpunit/tests/Story_Renderer.php [hunk 2]
```diff
@@ -76,12 +76,12 @@ public function test_replace_body_start_tag() {
 			echo 'Hello World';
 		};
 
-		add_action(	'web_stories_body_open', $function );
+		add_action( 'web_stories_body_open', $function );
```
Comment (barklund) [inline tests/phpunit/tests/Story_Renderer.php:79]: I have no idea what these changes are doing here. But the pre-commit complained, I let it auto-fix, and here we are.

Comment (swissspidy) [reply tests/phpunit/tests/Story_Renderer.php:79]: Odd that this wasn‘t caught in my PR 🤔

But the change is fine.

**Comment by spacedmonkey** [comment]
Weird thing I noticed, I couldn't undo gradient changes. As in, I made 5 points in the gradient when I pressed undo, all the changes were reverted. Is this expected?

**Comment by spacedmonkey** [comment]
I noticed that the arrow keys do not work, as it doesn't matter what directory, the marker always goes left.

**Comment by spacedmonkey** [comment]
I am also going to post my thoughts I posted on the gutenberg gradient picker as they are very similar in design and functionality. https://github.com/WordPress/gutenberg/pull/17603#issuecomment-538122253

**Comment by spacedmonkey** [comment]
Should we add more testing to this? Maybe some basic e2e or intergration tests?

File: assets/src/edit-story/components/colorPicker/useKeyDeleteStop.js [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+
+/**
+ * Internal dependencies
+ */
+import { useKeyDownEffect } from '../keyboard';
+
+function useKeyDeleteStop(ref, onDelete) {
+  useKeyDownEffect(
+    ref,
+    ['del', 'backspace'],
```
Comment (swissspidy) [inline assets/src/edit-story/components/colorPicker/useKeyDeleteStop.js]: Don't we support `delete` as a special case for this? See `keys.js`.

Comment (barklund) [reply assets/src/edit-story/components/colorPicker/useKeyDeleteStop.js]: Ah yes, fixed in bc05c519.

Comment (barklund) [reply assets/src/edit-story/components/colorPicker/useKeyDeleteStop.js]: Ah yes, fixed in bc05c519.

File: assets/src/edit-story/components/colorPicker/useKeyMoveStop.js [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+ */
+
+/**
+ * Internal dependencies
+ */
+import { useKeyDownEffect } from '../keyboard';
+
+function useKeyMoveStop(ref, onMove) {
+  useKeyDownEffect(ref, 'left', () => onMove(-0.01), [onMove]);
+  useKeyDownEffect(ref, 'right', () => onMove(0.01), [onMove]);
```
Comment (barklund) [inline assets/src/edit-story/components/colorPicker/useKeyMoveStop.js:24]: Fixing bug in 710b51c0 as [mentioned by @spacedmonkey](https://github.com/google/web-stories-wp/pull/497#issuecomment-596438886) (bug obviously introduced in [this refactor](https://github.com/google/web-stories-wp/pull/497/commits/3391b65d58ad518c7003c8a315513305a9af81d4#diff-1c08caaa370a6aa0afb916eda8c427ab))

File: assets/src/edit-story/components/colorPicker/useKeyAddStop.js [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+    : stops[currentStopIndex];
+  const postStop = isLastStopFocussed
+    ? stops[currentStopIndex]
+    : stops[currentStopIndex + 1];
+  return preStop.position + (postStop.position - preStop.position) / 2;
+}
+
+function useKeyAddStop(ref, onAdd, stops, currentStopIndex) {
+  const newStopPosition = getNewStopPosition(stops, currentStopIndex);
+  useKeyDownEffect(ref, 'enter', () => onAdd(newStopPosition), [
```
Comment (barklund) [inline assets/src/edit-story/components/colorPicker/useKeyAddStop.js]: When any gradient stop is focused, `'enter'` creates a new gradient stop. Regarding where the stop is added, see source for `getNewStopPosition` above. This addresses [one of @spacedmonkey's comments made externally](https://github.com/WordPress/gutenberg/pull/17603#issuecomment-538122253).

How do we communicate this shortcut best?

**Comment by dvoytenko** [comment]
Since I'm out most of day today, LGTM on non-colorPicker package. Please wait for @swissspidy for colorPicker review.

**Comment by barklund** [comment]
Replying to various comments by @spacedmonkey:

> Weird thing I noticed, I couldn't undo gradient changes. As in, I made 5 points in the gradient when I pressed undo, all the changes were reverted. Is this expected?

Works for me. I can undo and redo each separate gradient change just fine - done with mouse or keyboard. Basically, if the actual page layout updates when you change any individual thing, each individual change will **always** create an entry in the history. It simply _can't **not** happen_.

> I noticed that the arrow keys do not work, as it doesn't matter what directory, the marker always goes left.

[Well spotted](https://github.com/google/web-stories-wp/pull/497#discussion_r389826363).

> I am also going to post my thoughts I posted on the gutenberg gradient picker as they are very similar in design and functionality. [WordPress/gutenberg#17603 (comment)](https://github.com/WordPress/gutenberg/pull/17603#issuecomment-538122253)

You have many separate comments there, trying to address them individually:

* _Removing a control point should be easy_: need UX for this. Currently you can [drag a stop 30px off the gradient](https://github.com/google/web-stories-wp/pull/497/files#diff-a966e8597efe16433b5b0046d1dc96f9R42-R47) or [press delete/backspace when focused](https://github.com/google/web-stories-wp/pull/497/files#diff-1c93538114e0adb55be54027d8e32614R23-R32). If we want to make it more explicit, UX is needed.
* _Overlaying control points are currently possible_: Yes. And the result UX is a bit tricky. Potential mitigation should happen both when [adding](https://github.com/google/web-stories-wp/pull/497/files#diff-f66356393c2a059879321d62f9ca2ffdR118-R138) and [moving stops](https://github.com/google/web-stories-wp/pull/497/files#diff-f66356393c2a059879321d62f9ca2ffdR139-R168). UX decision needed.
* _I don't think a mark should able to be move pass the next order in order_: Why not? Works great for me - [re-ordering stops happen here](https://github.com/google/web-stories-wp/pull/497/files#diff-f66356393c2a059879321d62f9ca2ffdR155-R156). UX decision needed.
* _Should there be a max number of control points?_ : If need be, [it could very easily be added](https://github.com/google/web-stories-wp/pull/497/files#diff-f66356393c2a059879321d62f9ca2ffdR118-R122). UX decision needed.
* _Delete gradient stop by keyboard_: [Implemented here](https://github.com/google/web-stories-wp/pull/497/files#diff-1c93538114e0adb55be54027d8e32614R23-R32).
* _Move gradient stop by keyboard_: [Implemented here](https://github.com/google/web-stories-wp/pull/497#discussion_r389826363).
* _Add gradient stop by keyboard_: [Implemented and addressed here](https://github.com/google/web-stories-wp/pull/497#discussion_r389827910).

> Should we add more testing to this? Maybe some basic e2e or intergration tests?

I want to add a lot more testing to the color picker, but it's very time intensive. [I am working on it](https://github.com/google/web-stories-wp/pull/497/files#diff-8a375499c4540ddd717bc7df5ce5ade3), though. It'll be mostly integration testing for this complex component.

@spacedmonkey for further comments and discussion, please if at all possible make them inline, as it's _so_ much easier to have a conversation going in an inline thread rather than in disjoint comments on the main PR. I have linked to all relevant parts of the PR inline above. Thanks 😃

File: assets/src/edit-story/components/colorPicker/test/colorPicker.js [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+
+    const { closeButton, onClose } = arrange();
+    expect(document.body).not.toHaveFocus();
+
+    // press escape and wait for onClose to be invoked
+    await wait(() => {
+      const promise = getResolvingPromise(onClose);
+      fireEvent.click(closeButton);
+      return promise;
+    });
```
Comment (barklund) [inline assets/src/edit-story/components/colorPicker/test/colorPicker.js]: This has a lot of duplicated logic, but my attempt at DRY'ing it up using `describe.each` resulted in a weird bug about duplicated renders. And also [KCD suggests not to nest anyway](https://kentcdodds.com/blog/avoid-nesting-when-youre-testing), though his examples are not at all related to this use-case.

Comment (barklund) [reply assets/src/edit-story/components/colorPicker/test/colorPicker.js]: Currently [debugging what's going on with the failing `each`](https://twitter.com/barklund/status/1237245630441050112).

**Review by swissspidy — APPROVED** [review]
Impressive work! 💪 

Edit: Just noticed [a small thing](https://github.com/google/web-stories-wp/pull/528/files#r390849753) while looking at another PR, but I'd say this should be handled in a new PR.

Let's get this beast in first!

**Pull request merged**